### PR TITLE
test/ci: support for clang asan/tsan in build and travis.

### PR DESCRIPTION
**Pull request by htuch**
* Add --config={clang-asan,clang-tsan} configuration option to Bazel CLI for when building with
  clang-5.0.

* Add clang-asan, which also checks some static initialization order stuff.

* Add ubsan to asan/clang-asan configs, the two play nicely together.

* Switch bazel.asan from gcc-4.9+asan to clang-5.0+clang-asan.

* Add bazel.tsan CI target, enabled in .travis.yml.

* Minor fixes to filesystem_impl_test and uuid_util_test that were necessary to make ASAN/UBSAN
  combo on clang-5.0 work.

MSAN is not included due to the issue discussed at https://github.com/lyft/envoy/issues/443.

### CODE REVIEW

File: bazel/README.md [hunk 1]
```diff
@@ -147,13 +147,32 @@ You can use the `-c <compilation_mode>` flag to control this, e.g.
...
 ```
-To build and run tests with the compiler's address sanitizer (ASAN) enabled:
+
+
+## Sanitizers
+
+To build and run tests with the gcc-4.9 compiler's [address sanitizer
+(ASAN)](https://github.com/google/sanitizers/wiki/AddressSanitizer) and
+[undefined behavior
+(UBSAN)](https://developers.redhat.com/blog/2014/10/16/gcc-undefined-behavior-sanitizer-ubsan) sanitizer enabled:
```
Comment (dnoe) [inline bazel/README.md]: Ah, thanks.  I've been meaning to ask you why we didn't have ubsan.

File: ci/do_ci.sh [hunk 2]
```diff
@@ -19,24 +19,35 @@ function bazel_release_binary_build() {
...
-  bazel --batch test ${BAZEL_TEST_OPTIONS} -c dbg --config=asan @envoy//test/... \
+  bazel --batch test ${BAZEL_TEST_OPTIONS} -c dbg --config=clang-asan @envoy//test/... \
+    //:echo2_integration_test
+  exit 0
+elif [[ "$1" == "bazel.tsan" ]]; then
+  setup_clang_toolchain
+  echo "bazel TSAN debug build with tests..."
+  cd "${ENVOY_CONSUMER_SRCDIR}"
+  echo "Building and testing..."
+  bazel --batch test ${BAZEL_TEST_OPTIONS} -c dbg --config=tsan @envoy//test/... \
```
Comment (dnoe) [inline ci/do_ci.sh]: Should this be `--config=clang-tsan` ?

Comment (htuch) [inline ci/do_ci.sh]: Ah yeah, snafu. It does actually pass with `clang-tsan`, will update.

File: tools/bazel.rc [hunk 3]
```diff
@@ -2,10 +2,36 @@
...
+# Basic ASAN/UBSAN that works for gcc-4.9
+build:asan --copt -fsanitize=address,undefined
+build:asan --linkopt -fsanitize=address,undefined
 build:asan --linkopt -ldl
 build:asan --define tcmalloc=disabled
 build:asan --build_tag_filters=-no_asan
 build:asan --test_tag_filters=-no_asan
 build:asan --define signal_trace=disabled
+
+# Clang 5.0 ASAN
```
Comment (dnoe) [inline tools/bazel.rc]: Any reason why we don't use `-fsanitize=undefined` here too?

Comment (htuch) [inline tools/bazel.rc]: Too many  unaligned pointer issues. We can fix these later if we care (certainly a performance advantage and more interesting as we consider running on more than just x86), I'll add a TODO.

File: source/common/filesystem/filesystem_impl.cc [hunk 4]
```diff
@@ -31,7 +31,9 @@ bool fileExists(const std::string& path) {
 bool directoryExists(const std::string& path) {
   DIR* dir = opendir(path.c_str());
   bool dirExists = nullptr != dir;
-  closedir(dir);
+  if (dirExists) {
```
Comment (mattklein123) [inline source/common/filesystem/filesystem_impl.cc]: nit: can you fix this to be dir_exists

File: tools/bazel.rc (hunk 3 above)
Comment (mattklein123) [inline tools/bazel.rc]: Out of curiosity where are the unaligned pointer issues?

Comment (htuch) [inline tools/bazel.rc]: https://github.com/lyft/envoy/issues/914

**Comment by htuch** [comment]
I'm going to split out the code changes to deal with UBSAN from this PR and do a full rebase, there are now enough code changes that it makes sense to have them considered in distinct PRs.

**Pull request merged**